rust/opendp_tooling/src/proof.rs

```diff
@@ -0,0 +1,161 @@
+use std::{collections::HashMap, env, ffi::OsStr, path::PathBuf};
+
+use darling::{Error, FromMeta, Result};
+use syn::{AttributeArgs, Item, Type, TypePath};
+
+/// Traverses the filesystem, starting at src_dir, looking for .tex files.
+/// If more than one file is discovered with the same name, the value becomes None
+pub fn find_proof_paths(
+    src_dir: &std::path::Path,
+) -> std::io::Result<HashMap<String, Option<String>>> {
+    let mut proof_paths = HashMap::new();
+    find_unique_file_names_with_extension(&mut proof_paths, &OsStr::new("tex"), src_dir, src_dir)?;
+    Ok(proof_paths)
+}
+
+/// Writes a collection of proof paths to {OUT_DIR}/proof_paths.json.
+pub fn write_proof_paths(proof_paths: &HashMap<String, Option<String>>) -> Result<()> {
+    std::fs::write(
+        get_out_dir()?.join("proof_paths.json"),
+        serde_json::to_string(proof_paths).map_err(Error::custom)?,
+    )
+    .map_err(Error::custom)
+}
+
+/// Load proof paths from {OUT_DIR}/proof_paths.json.
+/// Assumes the file was written in the build script.
+pub fn load_proof_paths() -> Result<HashMap<String, Option<String>>> {
+    serde_json::from_str(
+        &std::fs::read_to_string(get_out_dir()?.join("proof_paths.json")).map_err(Error::custom)?,
+    )
+    .map_err(Error::custom)
+}
+
+/// The inner function for find_proof_paths
+fn find_unique_file_names_with_extension(
+    matches: &mut HashMap<String, Option<String>>,
+    file_extension: &OsStr,
+    root_dir: &std::path::Path,
+    dir: &std::path::Path,
+) -> std::io::Result<()> {
+    if dir.is_dir() {
+        for entry in std::fs::read_dir(dir)? {
+            let path = entry?.path();
+            if path.is_dir() {
+                find_unique_file_names_with_extension(matches, file_extension, root_dir, &path)?;
+            } else {
+                if path.extension() != Some(file_extension) {
+                    continue;
+                }
+                if let Some(file_name) = path.file_stem() {
+                    matches
+                        .entry(file_name.to_string_lossy().to_string())
+                        .and_modify(|v| drop(v.take()))
```

Comment:
It's hard to see that this replaces the entry value with None, a comment would help.